File: Data Structures & Algorithms/trapping-rain-water/submission-1.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int n = height.size();
        vector<int> pre(n);
        pre[0] = height[0];
        vector<int> post(n);
        post[n-1] = height[n-1];
        for (int i = 1; i < n; i++) {
            pre[i] = max(pre[i-1], height[i-1]);
        }
        for (int i = n-2; i >= 0; i--) {
            post[i] = max(post[i+1], height[i+1]);
        }
        int ans = 0;
        for (int i = 0; i < n; i++) {
            ans += max(0,min(pre[i], post[i]) - height[i]);
        }
        return ans;
    }
};

File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-1.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        ans = 0
        i, j = 0, 0
        mp = {}
        currMax = 0
        while j < len(s):
            mp[s[j]] = 1 + mp.get(s[j], 0)
            currMax = max(currMax, mp[s[j]])
            if j-i+1 - currMax <= k:
                ans = max(ans, j-i+1)
            if j-i+1 - currMax > k:
                mp[s[i]] -= 1
                i += 1
            j += 1
        return ans

